src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -445,96 +445,16 @@ JRT_LEAF(jdouble, SharedRuntime::l2d(jlong x))
   return (jdouble)x;
 JRT_END
 
-// Reference implementation at src/java.base/share/classes/java/lang/Float.java:floatToFloat16
 JRT_LEAF(jshort, SharedRuntime::f2hf(jfloat  x))
-  union {jfloat f; jint i;} bits;
-  bits.f = x;
-  jint doppel = bits.i;
-  jshort sign_bit = (jshort) ((doppel & 0x80000000) >> 16);
-  if (g_isnan(x))
-    return (jshort)(sign_bit | 0x7c00 | (doppel & 0x007fe000) >> 13 | (doppel & 0x00001ff0) >> 4 | (doppel & 0x0000000f));
-
-  jfloat abs_f = (x >= 0.0f) ? x : (x * -1.0f);
-
-  // Overflow threshold is halffloat max value + 1/2 ulp
-  if (abs_f >= (65504.0f + 16.0f)) {
-    return (jshort)(sign_bit | 0x7c00); // Positive or negative infinity
-  }
-
-  // Smallest magnitude of Halffloat is 0x1.0p-24, half-way or smaller rounds to zero
-  if (abs_f <= (pow(2, -24) * 0.5f)) { // Covers float zeros and subnormals.
-    return sign_bit; // Positive or negative zero
-  }
-
-  jint exp = ((0x7f800000 & doppel) >> (24 - 1)) - 127;
-
-  // For binary16 subnormals, beside forcing exp to -15, retain
-  // the difference exp_delta = E_min - exp.  This is the excess
-  // shift value, in addition to 13, to be used in the
-  // computations below. Further the (hidden) msb with value 1
-  // in f must be involved as well
-  jint exp_delta = 0;
-  jint msb = 0x00000000;
-  if (exp < -14) {
-    exp_delta = -14 - exp;
-    exp = -15;
-    msb = 0x00800000;
-  }
-  jint f_signif_bits = ((doppel & 0x007fffff) | msb);
-
-  // Significand bits as if using rounding to zero
-  jshort signif_bits = (jshort)(f_signif_bits >> (13 + exp_delta));
-
-  jint lsb = f_signif_bits & (1 << (13 + exp_delta));
-  jint round  = f_signif_bits & (1 << (12 + exp_delta));
-  jint sticky = f_signif_bits & ((1 << (12 + exp_delta)) - 1);
-
-  if (round != 0 && ((lsb | sticky) != 0 )) {
-    signif_bits++;
-  }
-
-  return (jshort)(sign_bit | ( ((exp + 15) << 10) + signif_bits ) );
+  assert(StubRoutines::f2hf() != nullptr, "floatToFloat16 intrinsic is not supported on this platform");
+  typedef jshort (*f2hf_stub_t)(jfloat x);
+  return ((f2hf_stub_t)StubRoutines::f2hf())(x);
```

Comment:
What's the point of keeping the wrappers around? The stubs can be called directly, can't they?

Reply:
I wanted isolate function type cast and assert in one place.
BTW the comment in assert should be "the stub is not implemented on this platform".